packages/nbdime/src/common/mergeview.ts

```diff
@@ -143,563 +341,404 @@ function createNbdimeMergeView(
+      let decorationKey = this.getDecorationKey(sources);
+      effects.push(removeHighlightEffect.of({highlightType: 'chunk', decorationKey: decorationKey}));
+      effects.push(removeHighlightEffect.of({highlightType: 'conflict', decorationKey: decorationKey}));
+      effects.push(removeHighlightEffect.of({highlightType: 'start', decorationKey: decorationKey}));
+      effects.push(removeHighlightEffect.of({highlightType: 'end', decorationKey: decorationKey}));
+    }
+    if (editor !== this.baseEditorWidget.cm) {
+      effects.push(removeGutterMarkerEffect.of({type: 'all' }));
+      effects.push(removeLineChunkMappingEffect.of({type: 'picker'}));
+      effects.push(removeLineChunkMappingEffect.of({type: 'conflict'}));
     }
-    clearMarks(editor, arr, classes);
+    return effects;
   }
-}
-
-function isTextMarker(marker: Marker): marker is CodeMirror.TextMarker {
-  return 'clear' in marker;
-}
 
-function clearMarks(editor: CodeMirror.Editor, arr: Marker[], classes: DiffClasses) {
-  for (let i = arr.length - 1; i >= 0; --i) {
-    let mark = arr[i];
-    if (isTextMarker(mark)) {
-      mark.clear();
-      arr.splice(i, 1);
-    } else if ((mark as any).parent) {
-      editor.removeLineClass(mark, 'background', classes.chunk);
-      editor.removeLineClass(mark, 'background', classes.start);
-      editor.removeLineClass(mark, 'background', classes.end);
-      editor.removeLineClass(mark, 'background', CHUNK_CONFLICT_CLASS);
-      // Merge editor does not set a marker currently, so don't clear for it:
-      if (valueIn(classes.gutter, [leftClasses.gutter, rightClasses.gutter])) {
-        editor.setGutterMarker(mark, GUTTER_PICKER_CLASS, null);
-      } else {
-        editor.setGutterMarker(mark, GUTTER_CONFLICT_CLASS, null);
-      }
-      let line = editor.lineInfo(mark);
-      if (!line.bgClass || line.bgClass.length === 0) {
-        arr.splice(i, 1);
+  private buildCharacterHighlighting(
+    editor: EditorView,
+    diffRanges: DiffRangePos[],
+    markType: DIFF_OP
+  ) {
+    let effects: StateEffect<unknown>[] = [];
+    let sources: ChunkSource[] = [];
+    if (markType === DIFF_OP.DIFF_INSERT || markType === DIFF_OP.DIFF_DELETE) {
+      let highlightType: string = markType === DIFF_OP.DIFF_DELETE ? 'deleted' : 'inserted';
+
+      for (let r of diffRanges) {
+        if (r.source !== undefined) {
+          sources.push(r.source);
+        }
+        let decorationKey = this.getDecorationKey(sources);
+        let startingOffset = posToOffset(editor.state.doc, {
+          line: r.from.line,
+          column: r.from.column
+        });
+        let endingOffset = posToOffset(editor.state.doc, {
+          line: r.to.line,
+          column: r.to.column
+        });
+        effects.push(addHighlightEffect.of({
+          from: startingOffset,
+          to: endingOffset,
+          highlightType: highlightType,
+          decorationKey: decorationKey
+        }));
       }
```

Comment:
All the local variables of this function should be const

Reply:
Thanks. This has been modified.